app/uk/gov/hmrc/mdgstub/controllers/MdgStubController.scala

```diff
@@ -16,76 +16,125 @@
 
 package uk.gov.hmrc.mdgstub.controllers
 
-import java.io.{ByteArrayInputStream, StringReader}
-
-import javax.inject.Singleton
+import java.io.StringReader
+import java.time.Instant
+import javax.inject.{Inject, Singleton}
 import org.xml.sax.SAXParseException
+import play.api.libs.concurrent.Promise
 import play.api.mvc._
 import uk.gov.hmrc.play.bootstrap.controller.BaseController
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
 import scala.util.{Failure, Success, Try}
 import scala.xml._
 
 @Singleton()
-class MdgStubController extends BaseController {
+class MdgStubController @Inject() (implicit ec: ExecutionContext) extends BaseController {
+
+  private val availabilityMode = raw"(\d{3}):(\d+):(\d+)".r
 
   def requestTransfer() = Action.async(parse.raw) { implicit request =>
 
-    val xml = new String(request.body.asBytes().get.toArray)
+    validateXml(request.body) match {
+      case Success(xml) =>
 
-    validXml(xml) match {
-      case Success(_) =>
-        if (checkIfSimulatedFailure(xml)) {
-          Future.successful(InternalServerError("Simulated failure"))
-        } else {
-          Future.successful(NoContent)
+        withActiveAvailabilityMode(xml) {
+          case Right(Some(AvailabilityMode(status,delay,_))) if (delay.length > 0) => Promise.timeout(Status(status.toInt), delay)
+          case Right(Some(AvailabilityMode(status,_,_)))     => Future.successful((Status(status.toInt)))
```

Comment:
What values would trigger this one?

Reply:
First case statement will match: "503:5:123456".
Second case statement will match: "503:0:123456".
i.e. there's a special case for when the delay is zero, because we don't want to call Promise.timeout() with a delay of zero. (As I found out the hard way ;-) ).